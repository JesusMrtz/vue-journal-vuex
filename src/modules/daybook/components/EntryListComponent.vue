<template>
  <div class="entry-list-container">
    <div class="px-2 pt-2">
      <input class="form-control" type="text" placeholder="Buscar entrada" v-model.trim="term">
    </div>
    <div class="mt-2 d-grid gap-2">
      <button type="button" class="btn btn-primary mx-3" @click="this.$router.push({ name: 'entry', params: { id: 'new' } })">
        <i class="fa fa-plus-circle"></i>
        Nueva entrada
      </button>
    </div>
    <div class="entry-scrollarea">
      <Entry v-for="entry in entriesByTerm" :key="entry.id" :entry="entry" />
    </div>
  </div>
</template>

<script>
import { defineAsyncComponent } from '@vue/runtime-core'
import { mapGetters } from 'vuex'


export default {
  components: {
    Entry: defineAsyncComponent(() => import('./EntryComponent'))
  },

  data() {
    return {
      term: ''
    }
  },

  computed: {
    ...mapGetters('journal', ['getEntriesByTerm']),
    entriesByTerm() {
      return this.getEntriesByTerm(this.term)
    }
  }
}
</script>

<style lang="scss" scoped>
.entry-list-container {
  border-right: 1px solid #2c3e50;
  height: calc(100vh - 56px);
}

.entry-scrollarea {
  height: calc(100vh - 115px);
  overflow-y: scroll;
}
</style>