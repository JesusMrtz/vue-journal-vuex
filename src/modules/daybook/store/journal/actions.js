/**
 * Son acciones que pueden ser asincronas que puede llamar una mutaci√≥n
 */
import journalAPI from "@/api/journalApi";


export async function loadEntries({ commit }) {
  try {
    const { data } = await journalAPI.get('/entries.json')
    const entries = []
  
    if ( !data ) commit('setEntries', [])
    
    for( let id of Object.keys(data) ) {
      entries.push({
        id,
        ...data[id]
      })
    }

    commit('setEntries', entries)
  } catch (error) {
    console.log(error)
  }
}

export async function updateEntry({ commit }, entry) {
  try {
    const { id, text, date, picture } = entry
    await journalAPI.put(`/entries/${id}.json`, ({ text, date, picture }))
   
    commit('updateEntry', {
      id,
      text,
      date,
      picture
    })
  } catch (error) {
    console.log(error)
  }
}

export async function createEntry({ commit }, entry) {
  try {
    const { data: { name } } = await journalAPI.post(`/entries.json`, entry)
    
    entry.id = name
    commit('addEntry', entry)

    return name;
  } catch (error) {
    console.log(error)
  }
}

export async function deleteEntry({ commit }, id) {
  try {
    await journalAPI.delete(`/entries/${id}.json`)
    commit('deleteEntry', id)
  } catch (error) {
    console.log(error)
  }
}