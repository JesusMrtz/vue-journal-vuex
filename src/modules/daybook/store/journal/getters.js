/**
 * Traer informaci√≥n del state que se puede traer del state,
 * Es parecido a las computed properties
 */
// export function myGetther( state ) {
//   return state;
// }