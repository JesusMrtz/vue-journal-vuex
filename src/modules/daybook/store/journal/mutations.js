/**
 * Son mÃ©todos sincronos que modifican el state
 */

export function setEntries(state, entries) {
  state.entries = [...state.entries, ...entries]
  state.isLoading = false
}

export function updateEntry(state, entry) {
  const idx = state.entries.findIndex(({ id }) => entry.id === id);
  state.entries[idx] = entry;
}

export function addEntry(state, entry) {
  state.entries.unshift(entry)
}

export function deleteEntry(state, id) {
  const deleteEntryById = state.entries.filter(entry => entry.id !== id)
  state.entries = deleteEntryById
}