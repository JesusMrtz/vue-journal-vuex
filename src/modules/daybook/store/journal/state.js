/**
 * el state es como la data de un compente en view que es reactivo
 */
export default () => ({
  isLoading: true,
  entries: [
    {
      id: new Date().getTime(),
      date: new Date().toDateString(),
      text: 'Adipisicing ullamco esse est voluptate mollit et ex est qui voluptate. Ad labore laborum ipsum minim aliqua id laboris ad velit ipsum cillum. Ad voluptate sint consectetur sit sint nulla ullamco veniam dolor laborum. Excepteur proident laborum aute qui Lorem do aute proident occaecat nulla cupidatat qui labore.',
      picture: null
    },
    {
      id: new Date().getTime() + 200,
      date: new Date().toDateString(),
      text: 'Ea nostrud commodo ullamco aute esse nisi aliquip dolor eu. Incididunt non culpa qui exercitation id nisi eu dolore do officia aute occaecat ut. Nisi magna excepteur sunt enim aliquip proident nostrud ad. Veniam sint enim ex occaecat dolore magna. Dolor pariatur ut quis pariatur laborum ex occaecat.',
      picture: null
    },
    {
      id: new Date().getTime() + 20000,
      date: new Date().toDateString(),
      text: 'Eu deserunt labore non enim deserunt ullamco. Excepteur tempor culpa eiusmod ipsum ea labore eiusmod do do deserunt sunt. Enim adipisicing reprehenderit magna consectetur velit qui in dolor et sit. Voluptate consequat labore nulla elit quis aliqua occaecat nostrud. Sint eu laboris aliquip ea adipisicing esse velit quis veniam nisi sunt mollit.',
      picture: null
    }
  ]
})