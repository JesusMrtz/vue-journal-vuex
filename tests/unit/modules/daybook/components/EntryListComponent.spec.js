import { createStore } from "vuex"
import { shallowMount } from "@vue/test-utils"
import { journalState } from "../../../mocks/test-journal-state"
import journal from '@/modules/daybook/store/journal'
import EntryListComponent from '@/modules/daybook/components/EntryListComponent'

const createVuexStore = (initialState) => 
createStore({
  modules: {
    journal: {
      ...journal,
      state: { ...initialState }
    }
  }
})

describe('Pruebas en el EntryList', () => {
  const mockRouter = {
    push: jest.fn()
  }
  const store = createVuexStore(journalState)
  let wrapper

  beforeEach(() => {
    jest.clearAllMocks()
    wrapper = shallowMount(EntryListComponent, {
      global: {
        mocks: {
          $router: mockRouter
        },
        plugins: [ store ]
      }
    })
  })

  it('Debe de llamar el getEntriesByTem sin tÃ©rminos y mostrar 2 entradas', async () => {
    expect(wrapper.findAll('entry-stub').length).toBe(2)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it ('Debe llar el getEntriesByTerm y filtrar las entradas', async() => {
    const input = wrapper.find('input')
    await input.setValue('segunda')
    
    expect(wrapper.findAll('entry-stub').length).toBe(1)
  })

  it('El boton de "Nuevo" debe de redireccionar a /new', async() => {
    await wrapper.find('button').trigger('click')

    expect(mockRouter.push).toHaveBeenCalledWith({ name: 'entry', params: {id: 'new'} })
  })
})