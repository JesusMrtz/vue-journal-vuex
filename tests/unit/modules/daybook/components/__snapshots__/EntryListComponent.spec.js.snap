// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el EntryList Debe de llamar el getEntriesByTem sin t√©rminos y mostrar 2 entradas 1`] = `
<div class="entry-list-container">
  <div class="px-2 pt-2"><input class="form-control" type="text" placeholder="Buscar entrada"></div>
  <div class="mt-2 d-grid gap-2"><button type="button" class="btn btn-primary mx-3"><i class="fa fa-plus-circle"></i> Nueva entrada </button></div>
  <div class="entry-scrollarea">
    <entry-stub entry="[object Object]"></entry-stub>
    <entry-stub entry="[object Object]"></entry-stub>
  </div>
</div>
`;
